src/llmq/quorums_instantsend.h

```diff
@@ -0,0 +1,149 @@
+     * to ixlocks.
+     */
+    std::unordered_map<uint256, CInstantXLockInfo, StaticSaltedHasher> finalInstantXLocks;
+    std::unordered_map<uint256, CInstantXLockInfo*, StaticSaltedHasher> txToInstantXLock;
+    std::unordered_map<COutPoint, CInstantXLockInfo*, SaltedOutpointHasher> inputToInstantXLock;
+    std::unordered_multimap<uint256, CInstantXLockInfo*, StaticSaltedHasher> blockToInstantXLocks;
+
+    const CBlockIndex* pindexLastChainLock{nullptr};
+
+    // Incoming and not verified yet
+    std::unordered_map<uint256, std::pair<NodeId, CInstantXLock>> pendingInstantXLocks;
+    bool hasScheduledProcessPending{false};
+
+public:
+    CInstantSendManager(CScheduler* _scheduler);
+    ~CInstantSendManager();
+
+    void RegisterAsRecoveredSigsListener();
+    void UnregisterAsRecoveredSigsListener();
+
+public:
+    bool ProcessTx(CNode* pfrom, const CTransaction& tx, CConnman& connman, const Consensus::Params& params);
+    bool CheckCanLock(const CTransaction& tx, bool printDebug, const Consensus::Params& params);
+    bool CheckCanLock(const COutPoint& outpoint, bool printDebug, const uint256* _txHash, CAmount* retValue, const Consensus::Params& params);
+    bool IsLocked(const uint256& txHash);
+    bool IsConflicted(const CTransaction& tx);
+    bool GetConflictingTx(const CTransaction& tx, uint256& retConflictTxHash);
+
+    virtual void HandleNewRecoveredSig(const CRecoveredSig& recoveredSig);
+    void HandleNewInputLockRecoveredSig(const CRecoveredSig& recoveredSig, const uint256& txid);
+    void HandleNewInstantXLockRecoveredSig(const CRecoveredSig& recoveredSig);
+
+    void TrySignInstantXLock(const CTransaction& tx);
+
+    void ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStream& vRecv, CConnman& connman);
+    void ProcessMessageInstantXLock(CNode* pfrom, const CInstantXLock& ixlock, CConnman& connman);
+    bool PreVerifyInstantXLock(NodeId nodeId, const CInstantXLock& ixlock, bool& retBan);
+    void ProcessPendingInstantXLocks();
+    void ProcessInstantXLock(NodeId from, const uint256& hash, const CInstantXLock& ixlock);
+    void UpdateWalletTransaction(const uint256& txid);
+
+    void SyncTransaction(const CTransaction &tx, const CBlockIndex *pindex, int posInBlock);
+    void NotifyChainLock(const CBlockIndex* pindex);
+    void UpdateIxLockMinedBlock(CInstantXLockInfo* ixlockInfo, const CBlockIndex* pindex);
+    void RemoveFinalIxLock(const uint256& hash);
+
+    void RemoveMempoolConflictsForLock(const uint256& hash, const CInstantXLock& ixlock);
+    void RetryLockMempoolTxs(const uint256& lockedParentTx);
+
+    bool AlreadyHave(const CInv& inv);
+    bool GetInstantXLockByHash(const uint256& hash, CInstantXLock& ret);
+};
+
+extern CInstantSendManager* quorumInstantSendManager;
+
+// The meaning of spork 2 has changed in v0.14. Before that, spork 2 was simply time based and either enabled or not
```

Comment:
This would be my preference too.

Reply:
For some reason I thought keeping number of sporks minimal was a desired thing :see_no_evil: 
I'll introduce a new spork now.

Reply:
Hmm... 62db023 is not what I meant exactly. To clarify: the idea was that spork2 would control the old system and spork20 would control the new one. We would turn spork2 off and spork20 on on migration and would later drop spork2 and all the legacy code completely in some future version. This way nodes recognizing spork2 only would know that this system is offline and they should not rely on it anymore.

Reply:
My thoughts in regard to compatibility was that old nodes don't have to know that a new system is in place, as they would simply stop receiving enough IS locks. Based on that, I assumed that the best solution would be to keep spork2 so that later when we remove the legacy system, everything looks like as it is today: spork2 contrils IS on/off and `instantlock` in RPC outputs tells if lock was successful.

Reply:
🤔